packages/teuchos/core/src/Teuchos_any.hpp

```diff
@@ -80,6 +84,66 @@
 
 namespace Teuchos {
 
+template<class T>
+struct is_comparable
+{
+    template<class X>
+    static auto test(int) -> decltype(std::declval<X>() == std::declval<X>(),
+                                      void(), std::true_type());
+    template<class X>
+    static auto test(...) -> std::false_type;
+    using type = decltype(test<T>(0));
+};
+
+template<class T>
+struct is_printable
+{
+    template<class X>
+    static auto test(int) -> decltype(std::declval<std::ostream&>() << std::declval<X>(),
+                                      void(), std::true_type());
+    template<class X>
+    static auto test(...) -> std::false_type;
+    using type = decltype(test<T>(0));
+};
+
+template <class T, class ok = typename is_comparable<T>::type>
+struct compare;
+
+template <class T>
+struct compare<T, std::false_type> {
+  bool operator()(T const&, T const&) const {
+    TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error,
```

Comment:
There are uses in which this is called, namely if one tries to compare equality of two `Teuchos::any` objects whose held type doesn't support it. I did not add a test that checks that these exceptions are called, I can do that as part of this PR.

Reply:
>  I did not add a test that checks that these exceptions are called, I can do that as part of this PR.

Thanks.  I think that would help to understand the context of this code as well.

What types would you hold in `Teuchos::any` (which requires a value-type object) that did not support operator==()?  I can understand value-types that don't support operator<<(std::print) but it is usually so easy to add one that I don't see why you would not add one.  Having debugging print levels is super useful.  What use case for `Teuchos::any` need to hold types that don't support these operators?  Is it some Panzer types?  But pretty much anything you put in a `Teuchos::ParameterList` object need to support these types I think (unless someone can give a good argument why they should not).

Reply:
Some example types are `std::set<T>`, `std::vector<T>`, and `Kokkos::View<T,...>`. None support printing, and `Kokkos::View<...>` doesn't support equality either. Note that this is _not_ using `Teuchos::ParameterList` at all, this is all in the context of `Teuchos::Reader` to store the results of parsed expressions. I can understand the appeal of having everything be printable, but with limited funding it is helpful not to force developers to add features their users may never exercise. Also, if I write a templated `Teuchos::Reader` that handles `std::set<T>` for any `T`, there is the design question of whether I should feel entitled to define the one and only `operator<<(std::ostream&, std::set<T> const&)`. What if someone else also tries to define it?

Reply:
I think the downside of these changes is that when people try to insert a type into a Teuchos::ParameterList that does not support the needed operators operator==() and operator<<() that instead of getting a compile-time error as they did before, they will now get a runtime error.  That is not the end of the world but compile-time errors are typically desired over runtime errors.  The former will be caught right away and the latter may not be caught until much later when it really hurts.

Could we put in extra compile-time checks in the Teuchos::ParameterList related class code that asserts that the underlying type has the needed operators?  I guess you could do that with a compile-time assert template and typedecl() calls like shown here, right?

Reply:
I don't personally have an issue with that restriction of `Teuchos::ParameterList`, so if no one else objects then yes I can put a `static_assert` for this into `Teuchos::ParameterList`. It is definitely doable.